File: PHP/DirectedEdgeTest.php
<?php

require_once('PHPUnit/Framework.php');
require_once('DirectedEdge.php');

class QueryTest extends PHPUnit_Framework_TestCase
{
    private $database;
    private $customer;
    private $product;

    protected function setUp()
    {
        $this->database = new DirectedEdgeDatabase(getenv('DIRECTEDEDGE_TEST_DB'),
                                                   getenv('DIRECTEDEDGE_TEST_PASS'));
        $this->database->import('../testdb.xml');

        $this->customer = new DirectedEdgeItem($this->database, 'customer1');
        $this->product = new DirectedEdgeItem($this->database, 'product1');
    }

    public function testLinks()
    {
        $this->assertEquals(15, count($this->customer->getLinks()));
        $this->assertArrayHasKey('product4', $this->customer->getLinks());

        $customer3 = new DirectedEdgeItem($this->database, 'customer3');
        $this->customer->linkTo($customer3, 10);

        $this->assertArrayHasKey('customer3', $this->customer->getLinks());
        $this->assertEquals(10, $this->customer->getWeightFor('customer3'));
    }

    /** 
     * @expectedException OutOfRangeException
     */
    public function testWeightUpperRange()
    {
        $this->customer->linkTo($this->product, 11);
    }

    /** 
     * @expectedException OutOfRangeException
     */
    public function testWeightLowerRange()
    {
        $this->customer->linkTo($this->product, -1);
    }

    public function testTags()
    {
        $this->assertEquals(1, count($this->customer->getTags()));
        $this->customer->addTag('foo');
        $this->assertContains('foo', $this->customer->getTags());
    }

    public function testRelated()
    {
        $this->assertEquals(5, count($this->product->getRelated(array(), array(maxResults => 5))));
        $this->assertContains('product21', $this->product->getRelated(array('product')));
    }

    public function testGroupRelated()
    {
        $results = $this->database->getGroupRelated(array('product1', 'product2'), array('product'));
        $this->assertEquals(20, count($results));

        $results = $this->database->getGroupRelated(array($this->product), array('product'));
        $this->assertEquals(20, count($results));

        $results = $this->database->getGroupRelated(array($this->product, 'product2'), array('product'));
        $this->assertEquals(20, count($results));
    }

    public function testRecommended()
    {
        $this->assertNotContains('product21', $this->customer->getRecommended(array('product')));
    }

    public function testProperties()
    {
        $this->customer->setProperty('foo', 'bar');
        $this->assertEquals('bar', $this->customer->getProperty('foo'));
        $this->customer->setProperty('baz', 'quux');
        $this->assertEquals('quux', $this->customer->getProperty('baz'));
        $this->customer->clearProperty('baz');
        $this->assertArrayNotHasKey('baz', $this->customer->getProperties());
        $this->assertArrayNotHasKey('quux', $this->customer->getProperties());
    }

    public function testSave()
    {
        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $foo->addTag('blah');
        $foo->save();

        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $this->assertContains('blah', $foo->getTags());
        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $foo->removeTag('blah');
        $foo->save();
        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $this->assertNotContains('blah', $foo->getTags());

        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $foo->setProperty('baz', 'quux');
        $foo->save();
        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $this->assertEquals('quux', $foo->getProperty('baz'));
        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $foo->setProperty('baz', 'bar');
        $foo->save();
        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $this->assertEquals('bar', $foo->getProperty('baz'));
        $foo->clearProperty('baz');
        $foo->save();
        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $this->assertArrayNotHasKey('baz', $foo->getProperties());

        $bar = new DirectedEdgeItem($this->database, 'Bar');
        $bar->linkTo($foo, 10);
        $bar->save();
        $bar = new DirectedEdgeItem($this->database, 'Bar');
        $links = $bar->getLinks();
        $this->assertEquals(10, $links['Foo']);
        $bar->unlinkFrom($foo);
        $bar->save();
        $bar = new DirectedEdgeItem($this->database, 'Bar');
        $this->assertArrayNotHasKey('Foo', $bar->getLinks());
    }

    public function testExport()
    {
        $exporter = new DirectedEdgeExporter('exported.xml');

        $foo = new DirectedEdgeItem($exporter->getDatabase(), 'Foo');
        $foo->addTag('blah');
        $foo->setProperty('baz', 'quux');
        $exporter->export($foo);

        $bar = new DirectedEdgeItem($exporter->getDatabase(), 'Bar');
        $bar->linkTo($foo, 5, 'magic');
        $exporter->export($bar);

        $exporter->finish();
    }

    public function testAdd()
    {
        $item = new DirectedEdgeItem($this->database, 'Asdf');
        $this->assertEquals(0, count($item->getTags()));
    }

    public function testLinkTypes()
    {
        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $bar = new DirectedEdgeItem($this->database, 'Bar');

        $foo->linkTo($bar, 0, 'thinger');

        $bar->save();
        $foo->save();

        $foo = new DirectedEdgeItem($this->database, 'Foo');
        $bar = new DirectedEdgeItem($this->database, 'Bar');

        $this->assertContains('thinger', $foo->getLinkTypes());
    }

    public function testDatabaseRelated()
    {
        $items = array('product1', 'product2', 'product3');
        $tags = array('product');
        $related = $this->database->getRelated(
            $items, $tags, array(threshold => 0.5 /* , countOnly => 'true' */));

        $results = $this->database->getRelated(array('product1', 'product2'), array('product'));
        $this->assertEquals(2, count($results));
        $this->assertEquals(20, count($results['product1']));
        $this->assertEquals(20, count($results['product2']));
        $this->assertEquals($this->product->getRelated(array('product')), $results['product1']);
    }
}

?>

File: Python/directed_edge_test.py
import directed_edge
import unittest
import os
import socket
import time

class QueryTest(unittest.TestCase):
    def setUp(self):
        self.database = directed_edge.Database(os.environ["DIRECTEDEDGE_TEST_DB"],
                                               os.environ["DIRECTEDEDGE_TEST_PASS"])
        self.database.import_from_file("../testdb.xml")
        self.customer = directed_edge.Item(self.database, "customer1")
        self.product = directed_edge.Item(self.database, "product1")

    def testLinks(self):
        self.assert_(len(self.customer.links()) == 15)
        self.assert_("product4" in self.customer.links())
        customer3 = directed_edge.Item(self.database, "customer3")
        self.customer.link_to(customer3, 10)
        self.assert_("customer3" in self.customer.links())
        self.assert_(self.customer.weight_for("customer3") == 10)
        self.assertRaises(Exception, self.customer.link_to, self.product, -1)
        self.assertRaises(Exception, self.customer.link_to, self.product, 11)

    def testTags(self):
        self.assert_(len(self.customer.tags) == 1)
        self.customer.add_tag("foo")
        self.assert_("foo" in self.customer.tags)

    def testRelated(self):
        self.assert_(len(self.product.related([], maxResults=5)) == 5)
        self.assert_(self.product.related([], popularity=0) !=
                     self.product.related([], popularity=1))
        self.assert_("product21" in self.product.related(["product"]))

    def testGroupRelated(self):
        self.assert_(len(self.database.group_related(["product1", "product2"], ["product"])) > 0)
        self.assert_(self.database.group_related(["product1"]) ==
                     self.product.related())

    def testProperties(self):
        self.customer["foo"] = "bar"
        self.assert_(self.customer.properties["foo"] == "bar")
        self.customer["baz"] = "quux"
        self.assert_(self.customer["baz"] == "quux")
        self.assert_(self.customer.get_property("baz") == "quux")
        self.customer.clear_property("baz")
        self.assert_(not self.customer.has_property("baz"))
        self.assert_(not self.customer.get_property("quux"))

    def testSave(self):
        item = lambda name: directed_edge.Item(self.database, name)

        foo = item("Foo")
        foo.add_tag("blah")
        foo.save()
        foo = item("Foo")
        self.assert_("blah" in foo.tags)
        foo = item("Foo")
        foo.remove_tag("blah")
        foo.save()
        foo = item("Foo")
        self.assert_("blah" not in foo.tags)

        foo = item("Foo")
        foo["baz"] = "quux"
        foo.save()
        foo = item("Foo")
        self.assert_(foo["baz"] == "quux")
        foo = item("Foo")
        foo["baz"] = "bar"
        foo.save()
        foo = item("Foo")
        self.assert_(foo["baz"] == "bar")
        foo.clear_property("baz")
        foo.save()
        foo = item("Foo")
        self.assert_("baz" not in foo.properties)        

        bar = item("Bar")
        bar.link_to(foo, 10)
        bar.save()
        bar = item("Bar")
        self.assert_(bar.links()["Foo"] == 10)
        bar.unlink_from(foo)
        bar.save()
        bar = item("Bar")
        self.assert_("Foo" not in bar.links())

    def testDestroy(self):
        customer = directed_edge.Item(self.database, "customer1")
        self.assert_(len(customer.links()) > 0)
        customer.destroy()
        customer = directed_edge.Item(self.database, "customer1")
        self.assert_(len(customer.links()) == 0)

    def testExport(self):
        exporter = directed_edge.Exporter("exported.xml")

        foo = directed_edge.Item(exporter.database, "Foo")
        foo.add_tag("blah")
        foo["baz"] = "quux"
        exporter.export(foo)

        bar = directed_edge.Item(exporter.database, "Bar")
        bar.link_to(foo, 5, "magic")
        exporter.export(bar)

        exporter.finish()

    def testAdd(self):
        exporter = directed_edge.Exporter(self.database)
        foo = directed_edge.Item(exporter.database, "Foo")
        foo.add_tag("blah")
        foo["baz"] = "quux"
        exporter.export(foo)
        exporter.finish()

        foo = directed_edge.Item(exporter.database, "Foo")
        self.assert_(foo["baz"] == "quux")

    def testNonexistant(self):
        item = directed_edge.Item(self.database, "Asdf")
        self.assert_(not item.tags)

    def testLinkTypes(self):
        foo = directed_edge.Item(self.database, "Foo")
        bar = directed_edge.Item(self.database, "Bar")

        foo.link_to(bar, 0, "thinger")

        bar.save()
        foo.save()

        foo = directed_edge.Item(self.database, "Foo")
        bar = directed_edge.Item(self.database, "Bar")

        self.assert_("thinger" in foo.link_types)

    def testCharacters(self):
        for id in [ ";@%&!", "foo/bar" ]:
            item = directed_edge.Item(self.database, id)
            item["foo"] = "bar"
            item.save()
            item = directed_edge.Item(self.database, id)
            self.assert_(item["foo"] == "bar")

    def testTimeout(self):
        if not "TEST_TIMEOUT" in os.environ:
            return

        database = directed_edge.Database("dummy", "dummy", "http",
                                          timeout = 5, host = "localhost:4567")

        start = time.time()
        timed_out = False
        timeout = 10
        item = directed_edge.Item(database, "dummy")
        try:
            item.tags()
        except socket.timeout:
            timed_out = True
            self.assert_(time.time() - start < timeout + 1)
        self.assert_(timed_out)

if __name__ == '__main__':
    unittest.main()
